src/tracers.jl

```diff
@@ -200,7 +200,18 @@ struct Dual{P<:Number,T<:Union{ConnectivityTracer,GradientTracer,HessianTracer}}
        AbstractTracer
     primal::P
     tracer::T
+
+    function Dual{P,T}(
+        primal::P, tracer::T
+    ) where {P<:Number,T<:Union{ConnectivityTracer,GradientTracer,HessianTracer}}
+        if P <: AbstractTracer
```

Comment:
wanna forbid `P <: Dual` too while we're at it?

Reply:
also this needs testing

Reply:
to check that an error is thrown, and prevent regressions

Reply:
Writing tests right now.
`P <: Dual` is tested since `Dual <: AbstractTracer`.

Reply:
I used the occasion to update our tests with our usual type abbreviations `C`, `G`, `H` and `S` for sets (instead of the previous `G`, which is taken for `GradientTracer`).

Reply:
good one, I wanted to do that ever since you moved the connectivity tests away from the gradient ones